### ISSUE TITLE
build failed

### ISSUE THREAD

**Issue by b4rthm** [bhuman/CompiledNN#3]
Could someone help me setup the project?
i tried making a main.cpp file with the code in the readme file. 
I extended the CMakeFile.txt with 
```
add_executable(main main.cpp)
target_link_libraries(main PRIVATE CompiledNN)
```
I build with
> mkdir _build && cd _build
> cmake .. -G "Unix Makefiles"
> cmake --build .

but i get the following error:

CMakeFiles/main.dir/main.cpp.o: In Funktion »NeuralNetwork::CompiledNN::apply() const«:
main.cpp:(.text._ZNK13NeuralNetwork10CompiledNN5applyEv[_ZNK13NeuralNetwork10CompiledNN5applyEv]+0x3c): Warnung: undefinierter Verweis auf »Assert::print(char const*, int, char const*, ...)«
main.cpp:(.text._ZNK13NeuralNetwork10CompiledNN5applyEv[_ZNK13NeuralNetwork10CompiledNN5applyEv]+0x41): Warnung: undefinierter Verweis auf »Assert::abort()«
collect2: error: ld returned 1 exit status
CMakeFiles/main.dir/build.make:90: recipe for target 'main' failed
make[2]: *** [main] Error 1
CMakeFiles/Makefile2:104: recipe for target 'CMakeFiles/main.dir/all' failed
make[1]: *** [CMakeFiles/main.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by ahasselbring** [comment]
Hi,
could it somehow be that CompiledNN was compiled with assertions disabled (Release-like) and main with assertions enabled (Debug-like)? That would perfectly explain the output, but seems very unlikely given your build commands. Could you check (using `make VERBOSE=1`) whether -DNDEBUG is passed to the compiler in 1. the CompiledNN target and 2. the main target?

**Comment by b4rthm** [comment]
I don't know.  Here is the output of `make VERBOSE=1` and I can't find the option -DNDEBUG . 

/snap/cmake/192/bin/cmake -S/home/barth/Workspace/CompiledNN -B/home/barth/Workspace/CompiledNN/_build --check-build-system CMakeFiles/Makefile.cmake 0
/snap/cmake/192/bin/cmake -E cmake_progress_start /home/barth/Workspace/CompiledNN/_build/CMakeFiles /home/barth/Workspace/CompiledNN/_build/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/home/barth/Workspace/CompiledNN/_build'
make -f CMakeFiles/CompiledNN.dir/build.make CMakeFiles/CompiledNN.dir/depend
make[2]: Entering directory '/home/barth/Workspace/CompiledNN/_build'
cd /home/barth/Workspace/CompiledNN/_build && /snap/cmake/192/bin/cmake -E cmake_depends "Unix Makefiles" /home/barth/Workspace/CompiledNN /home/barth/Workspace/CompiledNN /home/barth/Workspace/CompiledNN/_build /home/barth/Workspace/CompiledNN/_build /home/barth/Workspace/CompiledNN/_build/CMakeFiles/CompiledNN.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/barth/Workspace/CompiledNN/_build'
make -f CMakeFiles/CompiledNN.dir/build.make CMakeFiles/CompiledNN.dir/build
make[2]: Entering directory '/home/barth/Workspace/CompiledNN/_build'
make[2]: Nothing to be done for 'CMakeFiles/CompiledNN.dir/build'.
make[2]: Leaving directory '/home/barth/Workspace/CompiledNN/_build'
[ 97%] Built target CompiledNN
make -f CMakeFiles/main.dir/build.make CMakeFiles/main.dir/depend
make[2]: Entering directory '/home/barth/Workspace/CompiledNN/_build'
cd /home/barth/Workspace/CompiledNN/_build && /snap/cmake/192/bin/cmake -E cmake_depends "Unix Makefiles" /home/barth/Workspace/CompiledNN /home/barth/Workspace/CompiledNN /home/barth/Workspace/CompiledNN/_build /home/barth/Workspace/CompiledNN/_build /home/barth/Workspace/CompiledNN/_build/CMakeFiles/main.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/barth/Workspace/CompiledNN/_build'
make -f CMakeFiles/main.dir/build.make CMakeFiles/main.dir/build
make[2]: Entering directory '/home/barth/Workspace/CompiledNN/_build'
[ 98%] Linking CXX executable main
/snap/cmake/192/bin/cmake -E cmake_link_script CMakeFiles/main.dir/link.txt --verbose=1
/usr/bin/c++     CMakeFiles/main.dir/main.cpp.o  -o main  -Wl,-rpath,/home/barth/anaconda3/lib libCompiledNN.a -lpthread -lrt /home/barth/anaconda3/lib/libhdf5.so /usr/lib/x86_64-linux-gnu/librt.so -lpthread /home/barth/anaconda3/lib/libz.so /usr/lib/x86_64-linux-gnu/libdl.so -lm 
CMakeFiles/main.dir/main.cpp.o: In function `NeuralNetwork::CompiledNN::apply() const':
main.cpp:(.text._ZNK13NeuralNetwork10CompiledNN5applyEv[_ZNK13NeuralNetwork10CompiledNN5applyEv]+0x3c): undefined reference to `Assert::print(char const*, int, char const*, ...)'
main.cpp:(.text._ZNK13NeuralNetwork10CompiledNN5applyEv[_ZNK13NeuralNetwork10CompiledNN5applyEv]+0x41): undefined reference to `Assert::abort()'
collect2: error: ld returned 1 exit status
CMakeFiles/main.dir/build.make:90: recipe for target 'main' failed
make[2]: *** [main] Error 1
make[2]: Leaving directory '/home/barth/Workspace/CompiledNN/_build'
CMakeFiles/Makefile2:104: recipe for target 'CMakeFiles/main.dir/all' failed
make[1]: *** [CMakeFiles/main.dir/all] Error 2
make[1]: Leaving directory '/home/barth/Workspace/CompiledNN/_build'
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by ahasselbring** [comment]
OK, while it would have certainly been interesting to see the output of the build of CompiledNN, too (i.e. after recreating the build directory), there is something else that is interesting: CMake seems to use the HDF5 library from Anaconda (you probably do not have HDF5 installed otherwise on your system?). Due to the way CMake determines the compiler flags to use when linking with HDF5 it might be that a -DNDEBUG comes in from there into `HDF5_C_DEFINITIONS`, which are applied to the target `CompiledNN` but not to its consumers. You could verify this by either temporarily turning the `PRIVATE` in line 198 of `CMakeLists.txt` to `PUBLIC` or to check the output of `/home/barth/anaconda3/bin/h5cc -show` for -DNDEBUG.

Anyway, the solution should be to install `libhdf5-dev` (or whatever the HDF5 library is called in your system) and rebuild.

**Comment by b4rthm** [comment]
I checked and `libhdf5-dev` is already installed and the output of `/home/barth/anaconda3/bin/h5cc -show` contains -DNDEBUG. 

Here is the output of the build:
~/Workspace/CompiledNN/_build$ cmake --build .
Scanning dependencies of target CompiledNN
[  1%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/CompiledNN.cpp.o
[  2%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/Model.cpp.o
[  4%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/SimpleNN.cpp.o
[  5%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/CompiledNN/ActivationFunctions.cpp.o
[  7%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/CompiledNN/CompilationSettings.cpp.o
[  8%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/CompiledNN/Operations/Activation.cpp.o
[ 10%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/CompiledNN/Operations/Arithmetic.cpp.o
[ 11%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/CompiledNN/Operations/BatchNormalization.cpp.o
[ 12%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/CompiledNN/Operations/Concatenate.cpp.o
[ 14%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/CompiledNN/Operations/Conv2D.cpp.o
[ 15%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/CompiledNN/Operations/Cropping2D.cpp.o
[ 17%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/CompiledNN/Operations/DConv2D.cpp.o
[ 18%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/CompiledNN/Operations/Dense.cpp.o
[ 20%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/CompiledNN/Operations/GlobalPooling2D.cpp.o
[ 21%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/CompiledNN/Operations/Im2Col2D.cpp.o
[ 22%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/CompiledNN/Operations/Pooling2D.cpp.o
[ 24%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/CompiledNN/Operations/Softmax.cpp.o
[ 25%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/CompiledNN/Operations/UInt8Input.cpp.o
[ 27%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/CompiledNN/Operations/UpSampling2D.cpp.o
[ 28%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/CompiledNN/Operations/ZeroPadding2D.cpp.o
[ 30%] Building CXX object CMakeFiles/CompiledNN.dir/Src/CompiledNN/CompiledNN/Util/ExpApprox.cpp.o
[ 31%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/arch.cpp.o
[ 32%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/assembler.cpp.o
[ 34%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/builder.cpp.o
[ 35%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/callconv.cpp.o
[ 37%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/codeholder.cpp.o
[ 38%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/compiler.cpp.o
[ 40%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/constpool.cpp.o
[ 41%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/cpuinfo.cpp.o
[ 42%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/emitter.cpp.o
[ 44%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/func.cpp.o
[ 45%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/globals.cpp.o
[ 47%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/inst.cpp.o
[ 48%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/jitallocator.cpp.o
[ 50%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/jitruntime.cpp.o
[ 51%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/logging.cpp.o
[ 52%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/operand.cpp.o
[ 54%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/osutils.cpp.o
[ 55%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/ralocal.cpp.o
[ 57%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/rapass.cpp.o
[ 58%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/rastack.cpp.o
[ 60%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/string.cpp.o
[ 61%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/support.cpp.o
[ 62%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/target.cpp.o
[ 64%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/type.cpp.o
[ 65%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/virtmem.cpp.o
[ 67%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/zone.cpp.o
[ 68%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/zonehash.cpp.o
[ 70%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/zonelist.cpp.o
[ 71%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/zonestack.cpp.o
[ 72%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/zonetree.cpp.o
[ 74%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/core/zonevector.cpp.o
[ 75%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/x86/x86assembler.cpp.o
[ 77%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/x86/x86builder.cpp.o
[ 78%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/x86/x86callconv.cpp.o
[ 80%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/x86/x86compiler.cpp.o
[ 81%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/x86/x86features.cpp.o
[ 82%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/x86/x86instapi.cpp.o
[ 84%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/x86/x86instdb.cpp.o
[ 85%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/x86/x86internal.cpp.o
[ 87%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/x86/x86logging.cpp.o
[ 88%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/x86/x86operand.cpp.o
[ 90%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/asmjit/src/asmjit/x86/x86rapass.cpp.o
[ 91%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/B-Human/Platform/BHAssert.cpp.o
[ 92%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/B-Human/Tools/Streams/InOut.cpp.o
[ 94%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/B-Human/Tools/Streams/InStreams.cpp.o
[ 95%] Building CXX object CMakeFiles/CompiledNN.dir/3rdParty/B-Human/Tools/Streams/SimpleMap.cpp.o
[ 97%] Linking CXX static library libCompiledNN.a
[ 97%] Built target CompiledNN
Scanning dependencies of target main
[ 98%] Building CXX object CMakeFiles/main.dir/main.cpp.o
[100%] Linking CXX executable main
CMakeFiles/main.dir/main.cpp.o: In function `NeuralNetwork::CompiledNN::apply() const':
main.cpp:(.text._ZNK13NeuralNetwork10CompiledNN5applyEv[_ZNK13NeuralNetwork10CompiledNN5applyEv]+0x3c): undefined reference to `Assert::print(char const*, int, char const*, ...)'
main.cpp:(.text._ZNK13NeuralNetwork10CompiledNN5applyEv[_ZNK13NeuralNetwork10CompiledNN5applyEv]+0x41): undefined reference to `Assert::abort()'
collect2: error: ld returned 1 exit status
CMakeFiles/main.dir/build.make:90: recipe for target 'main' failed
make[2]: *** [main] Error 1
CMakeFiles/Makefile2:104: recipe for target 'CMakeFiles/main.dir/all' failed
make[1]: *** [CMakeFiles/main.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by ahasselbring** [comment]
So CMake is choosing the Anaconda-HDF5 by default. This can probably be avoided by passing `-DHDF5_ROOT=/usr` to CMake (-> `cmake -DHDF5_ROOT=/usr ..`).

**Comment by b4rthm** [comment]
The build finally runs through, that solved the problem.
Thank you very much.
